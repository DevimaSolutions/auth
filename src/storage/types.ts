export interface IStorage {
  setItem<T>(key: string, data: T): void;
  getItem<T>(key: string): T;
  remove(key: string): void;
  multiSet<T extends {}>(data: T): void;
  multiRemove(keys: string[]): void;
}
