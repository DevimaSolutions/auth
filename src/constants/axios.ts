export const authRetryAxiosRequestConfigFlag = '@ds/auth/isRetryRequest';
