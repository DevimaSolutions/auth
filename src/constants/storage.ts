export const defaultAuthStorageKeys = {
  accessToken: '@ds/auth/access',
  refreshToken: '@ds/auth/refresh',
  user: '@ds/auth/user',
} as const;
