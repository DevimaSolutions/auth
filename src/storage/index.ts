export { default } from './storage';

export type { IStorage } from './storage.types';
