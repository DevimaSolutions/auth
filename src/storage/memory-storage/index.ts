export { default } from './memory-storage';
