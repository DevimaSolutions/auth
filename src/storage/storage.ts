import AsyncStorage from '@react-native-async-storage/async-storage';

import type { IStorage } from './storage.types';

export default class Storage implements IStorage {
  private _normalizeData<T>(data: T): string {
    return typeof data === 'string' ? data : JSON.stringify(data);
  }

  private _denormalizeData<T>(data: string): T {
    try {
      return JSON.parse(data);
    } catch {
      throw new Error('Value is not an object');
    }
  }

  remove = AsyncStorage.removeItem;
  multiRemove = AsyncStorage.multiRemove;

  async setItem<T>(key: string, data: T): Promise<void> {
    AsyncStorage.setItem(key, this._normalizeData(data));
  }

  async setString(key: string, data: string): Promise<void> {
    AsyncStorage.setItem(key, data);
  }

  async getItem<T>(key: string): Promise<T> {
    const value = await AsyncStorage.getItem(key);

    if (!value) {
      throw new Error('Value not set');
    }

    return this._denormalizeData<T>(value);
  }

  async getString(key: string): Promise<string> {
    const value = await AsyncStorage.getItem(key);

    if (!value) {
      throw new Error('Value not set');
    }

    return value;
  }

  async multiSet<T extends Record<string, number | string | object>>(
    data: T
  ): Promise<void> {
    const entries = Object.entries(data).map(([key, value]) => [
      key,
      this._normalizeData(value),
    ]);
    await AsyncStorage.multiSet(entries);
  }
}
