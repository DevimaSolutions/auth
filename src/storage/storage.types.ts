export interface IStorage {
  setItem<T>(key: string, data: T): Promise<void>;
  setString(key: string, data: string): Promise<void>;
  getItem<T>(key: string): Promise<T>;
  getString(key: string): Promise<string>;
  remove(key: string): Promise<void>;
  multiSet<T extends Record<string, number | string | object>>(
    data: T
  ): Promise<void>;
  multiRemove(keys: string[]): Promise<void>;
}
