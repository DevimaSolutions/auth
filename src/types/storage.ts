import type { defaultAuthStorageKeys } from '../constants/storage';

export interface IAuthStorageKeys {
  accessToken: string;
  refreshToken: string;
  user: string;
}

export type IDefaultAuthStorageKeys = typeof defaultAuthStorageKeys;
