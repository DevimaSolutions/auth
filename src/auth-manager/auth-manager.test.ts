import axios from 'axios';
import MockAdapter from 'axios-mock-adapter';

import { MemoryStorage } from '../storage';

import AuthManager from './auth-manager';

import type { IAuthOptions, ISignedInOptions } from '../types';

const getAxiosMock = () => {
  const instance = axios.create();
  const mock = new MockAdapter(instance, { delayResponse: 10 });
  mock.onGet('/sign-in').reply(200, { token: 'valid-token' });
  mock.onGet('/refresh').reply(200, { token: 'valid-token' });
  mock.onGet('/user').reply(200, { id: 1 });
  return instance;
};

const getOptions = (): IAuthOptions<{ id: number }, { token: string }> => ({
  axiosInstance: getAxiosMock(),
  signIn: jest.fn(async () => getAxiosMock().get('/sign-in')),
  signOut: jest.fn(async () => {}),
  refreshToken: jest.fn(async () => getAxiosMock().get('/refresh')),
  getUser: jest.fn(async () => getAxiosMock().get('/user')),
  storage: new MemoryStorage(),
});

const signInOptions: ISignedInOptions<true, { id: number }> = {
  isSignedIn: true,
  accessToken: 'test-access',
  refreshToken: 'test-refresh',
  user: { id: 1 },
};

describe('AuthManager', () => {
  test('Can create auth manager', () => {
    expect(() => {
      new AuthManager(getOptions());
    }).not.toThrowError();
  });
  test('Can get options', () => {
    const manager = new AuthManager(getOptions());

    const actualResult = manager.options;

    expect(actualResult).toMatchSnapshot();
  });
  test('Can get axios instance', () => {
    const options = getOptions();
    const manager = new AuthManager(options);

    const actualResult = manager.axios;

    expect(actualResult).toBe(options.axiosInstance);
  });
  test('Should sign in on init', () => {
    const manager = new AuthManager(getOptions(), signInOptions);
    expect(manager.isSignedIn()).toBeTruthy();
  });
  test('Can get empty auth header when unauthorized', () => {
    const manager = new AuthManager(getOptions());
    expect(manager.getAuthorizationHeader()).toBeNull();
  });
  test('Can get auth header when authorized', () => {
    const manager = new AuthManager(getOptions(), signInOptions);
    expect(manager.getAuthorizationHeader()).toBe(`Bearer ${signInOptions.accessToken}`);
  });
  test('Can get empty user when unauthorized', () => {
    const manager = new AuthManager(getOptions());
    expect(manager.getUser()).toBeNull();
  });
  test('Can get user when authorized', () => {
    const manager = new AuthManager(getOptions(), signInOptions);
    expect(manager.getUser()).toEqual({ id: 1 });
  });
  test('Can get empty refresh token when unauthorized', () => {
    const manager = new AuthManager(getOptions());
    expect(manager.getRefreshToken()).toBeNull();
  });
  test('Can get refresh token when authorized', () => {
    const manager = new AuthManager(getOptions(), signInOptions);
    expect(manager.getRefreshToken()).toBe(signInOptions.refreshToken);
  });
});
