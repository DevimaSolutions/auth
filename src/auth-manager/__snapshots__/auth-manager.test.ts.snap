// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthManager Can get options 1`] = `
{
  "axiosInstance": [Function],
  "buildAuthorizationHeader": [Function],
  "getUser": [MockFunction],
  "refreshToken": [MockFunction],
  "signIn": [MockFunction],
  "signOut": [MockFunction],
  "storage": MemoryStorage {
    "_data": Map {},
    "getItem": [Function],
    "multiRemove": [Function],
    "multiSet": [Function],
    "remove": [Function],
    "setItem": [Function],
  },
  "storageKeys": {
    "accessToken": "@ds/auth/access",
    "refreshToken": "@ds/auth/refresh",
    "user": "@ds/auth/user",
  },
}
`;
